tmpl/blog/functoria.md

```diff
@@ -0,0 +1,153 @@
+
+- Set a good default directly in the `config.ml`
+- Provide a value at configure time, if you are already aware of deployment conditions.
+- Provide a value at startup time, for last minute changes.
+
+All of this is now possible using **keys**. A key is composed of :
+- _name_ : The name of the value in the program.
+- _description_ : How it should be displayed/serialized.
+- _stage_ : Is the key available only at runtime, at configure time or both ?
+- _documentation_ : It is not optional so you should really write it.
+
+Imagine we are building a multilingual unikernel and we want to pass the
+default language as a parameter. We will use a simple string, so we can use the
+predefined description `Key.Desc.string`. We want to be able to define it both
+at configure and run time, so we use the stage `` `Both``. This gives us the following code:
+
+```
+let lang_key =
+  let doc = Key.Doc.create
+      ~doc:"The default language for the unikernel." [ "l" ; "lang" ]
+  in
+  Key.create ~doc ~stage:`Both ~default:"en" "language" Key.Desc.string
+```
+
+Here, We defined both a long option `--lang` and a short one `-l` (the format is similar to the one used by [Cmdliner][cmdliner]).
+In the unikernel, the value is retrieved with `Bootvar_gen.language ()`.
+
+The option is also documented in the `--help` option for both `mirage configure` (at configure time) and `./my_unikernel` (at startup time).
+
+```
+       -l VAL, --lang=VAL (absent=en)
+           The default language for the unikernel.
+```
+
+[cmdliner]: http://erratique.ch/software/cmdliner
+
+A simple example of a unikernel with a key is available in [mirage-skeleton][] in the `hello` directory.
+
+### Switching implementation
+
+We can do much more with keys: we can use them to switch implementation at configure time.
+To illustrate, let us take the example of a dynamic storage: We want to choose between a block device and a crunch device with a command line option.
+In order to do that, we must first define a boolean key:
+
+```
+let fat_key =
+  let doc = Key.Doc.create
+      ~doc:"Use a fat device if true, crunch otherwise." [ "fat" ]
+  in
+  Key.create ~doc ~stage:`Configure ~default:false "fat" Key.Desc.bool
+```
+
+We can use the `if_impl` combinator to choose between two devices depending on the value of the key.
+
+```
+let dynamic_storage =
+  if_impl (Key.value fat_key)
+    (kv_ro_of_fs my_fat_device)
+    (my_crunch_device)
+```
+
+We can now use this device as a normal storage device of type `kv_ro impl`! The key is also documented in `mirage configure --help`:
+
+```
+       --fat=VAL (absent=false)
+           Use a fat device if true, crunch otherwise.
+```
+
+It is also possible to compute on keys before giving them to `if_impl`, combining multiple keys in order to compute a value, and so on. The documentation for the API is located in the `Mirage.Key` module and various examples are available in [mirage][] and [mirage-skeleton][].
```

Comment:
Yes, once it has an online documentation. For now 1) It's not merged 2) regular ocamldoc doesn't work very well with includes, making it not that useful. I would prefer having codoc ...